internal/controller/postgrescluster/postgres.go

```diff
@@ -628,6 +635,71 @@ func (r *Reconciler) reconcilePostgresDataVolume(
 	return pvc, err
 }
 
+// determineVolumeSize compares the potential sizes from the instance spec, status
+// and limit and sets the appropriate current value.
+func (r *Reconciler) determineVolumeSize(ctx context.Context, cluster *v1beta1.PostgresCluster,
+	pvc *corev1.PersistentVolumeClaim, instanceName, instanceSpecName string) {
+	log := logging.FromContext(ctx)
+
+	// Capture the largest pgData volume size currently defined for a given instance set.
+	volumeRequestSize := pvc.Spec.Resources.Requests.Storage().Value()
+
+	// If the request value is greater than the set limit, use the limit and issue
+	// a warning event. A limit of 0 is ignorned.
+	if !pvc.Spec.Resources.Limits.Storage().IsZero() &&
+		pvc.Spec.Resources.Requests.Storage().Value() > pvc.Spec.Resources.Limits.Storage().Value() {
+		r.Recorder.Eventf(cluster, corev1.EventTypeWarning, "VolumeRequestOverLimit",
+			"pgData volume request (%v) for %s/%s is greater than set limit (%v). Limit value will be used.",
+			pvc.Spec.Resources.Requests.Storage(), cluster.Name, instanceSpecName, pvc.Spec.Resources.Limits.Storage())
+
+		pvc.Spec.Resources.Requests = corev1.ResourceList{
+			corev1.ResourceStorage: *resource.NewQuantity(pvc.Spec.Resources.Limits.Storage().Value(), resource.BinarySI),
+		}
+		// Otherwise, if the limit is not set or the feature gate is not enabled, do not autogrow.
+	} else if !pvc.Spec.Resources.Limits.Storage().IsZero() && util.DefaultMutableFeatureGate.Enabled(util.AutoGrowVolumes) {
```

Comment:
This seems to be the definition of when auto-grow applies:
```
 !pvc.Spec.Resources.Limits.Storage().IsZero() && util.DefaultMutableFeatureGate.Enabled(util.AutoGrowVolumes) {
```
A function called something like `autogrowPermitted` might be nice to keep tabs on that definition.

Reply:
I thought about that, but in `observeInstances` I only check for the feature gate (because limits can vary per instance set). For now, I'm inclined to leave this as is, but might be worth talking through. Another thing worth noting is that the limits equal to zero (e.g. `0Gi`) will not trigger autogrow despite being acceptable in the manifest. I think this is a reasonable approach, but happy to hear your thoughts on it.